experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,697 @@
+      result->data = resData;
+      muliDelayBuffer.exec(isReset, joinValid, oehbReady, buffValid);
+      muliOEHB.exec(isReset, oehbDataIn, buffValid, result->ready, oehbReady,
+                    oehbDataOut, result->valid);
+    }
+  }
+
+private:
+  handshake::MulIOp op;
+  unsigned bitwidth;
+  Join muliJoin;
+  Mul4Stage muliMul4Stage;
+  DelayBuffer muliDelayBuffer;
+  OEHB muliOEHB;
+};
+
+class ReturnModel : public OpExecutionModel<handshake::ReturnOp> {
+public:
+  using OpExecutionModel<handshake::ReturnOp>::OpExecutionModel;
+  ReturnModel(handshake::ReturnOp returnOp)
+      : OpExecutionModel<handshake::ReturnOp>(returnOp), op(returnOp),
+        returnTEHB() {
+
+    auto temp =
+        cast<handshake::ChannelType>(returnOp.getInputs().front().getType());
+    datawidth = temp.getDataBitWidth();
+  }
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+    auto *ins = getState<ChannelConsumerRW>(rws[{op.getInputs().front(), op}]);
+    auto *outs =
+        getState<ChannelProducerRW>(rws[{op.getOutputs().front(), op}]);
+    bool regEnable = false, regNotFull = false;
+    Any dataReg;
+
+    for (int i = 0; i < 10; ++i) {
+      returnTEHB.exec(isReset, ins->valid, outs->ready, regNotFull,
+                      outs->valid);
+      if (isReset) {
+        dataReg = 0;
+      } else if (regEnable) {
+        dataReg = ins->data;
+      }
+      if (regNotFull) {
+        outs->data = ins->data;
+      } else {
+        outs->data = dataReg;
+      }
+      ins->ready = regNotFull;
+    }
+  }
+
+private:
+  handshake::ReturnOp op;
+  unsigned datawidth;
+  TEHBDataless returnTEHB;
+};
+
+class EndModel : public OpExecutionModel<handshake::EndOp> {
+public:
+  using OpExecutionModel<handshake::EndOp>::OpExecutionModel;
+  EndModel(handshake::EndOp endOp)
+      : OpExecutionModel<handshake::EndOp>(endOp), op(endOp) {
+    auto temp =
+        cast<handshake::ChannelType>(endOp.getInputs().front().getType());
+    bitwidth = temp.getDataBitWidth();
+  }
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+
+    auto *ins = getState<ChannelConsumerRW>(rws[{op.getInputs().front(), op}]);
+    auto *outs =
+        getState<ChannelProducerRW>(rws[{op.getReturnValues().front(), op}]);
+
+    ins->ready = ins->valid && outs->ready;
+    outs->valid = ins->valid;
+    outs->data = ins->data;
+  }
+
+private:
+  handshake::EndOp op;
+  unsigned bitwidth;
+};
+
+//===----------------------------------------------------------------------===//
+// Simulator
+//===----------------------------------------------------------------------===//
+
+class Simulator {
+public:
+  Simulator(handshake::FuncOp funcOp) : funcOp(funcOp) {
+    // Iterate through all the values of the circuit
+    for (BlockArgument arg : funcOp.getArguments()) {
+      // get the consumer op of the value
+      auto *user = *arg.getUsers().begin();
```

Comment:
But in fact, if the value's allowed to have more than 2 users (the consumer and producer), it breaks all the logic with valueStates, because Value is not a unique ID anymore :))
I expected that having more than 1 value.users() means that it's a copy of one channel/control between the same consumer and producer